crypto/signatures.go

```diff
@@ -69,8 +72,47 @@ func VerifyHash(data Hash, pk PublicKey, sig Signature) error {
 	return nil
 }
 
+// WriteSignedObject writes a length-prefixed object followed by its signature.
+func WriteSignedObject(w io.Writer, obj interface{}, sk SecretKey) error {
+	encObj := encoding.Marshal(obj)
+	sig, _ := SignHash(HashBytes(encObj), sk) // no error possible
+	return encoding.NewEncoder(w).EncodeAll(encObj, sig)
```

Comment:
The length prefix is necessary to know the number of bytes in the encoded object. This is why `encoding.WriteObject` and `encoding.ReadObject` use a length prefix. Almost every object sent over the wire is sent in this format. Not only does this allow you to reject large objects without having to read them, it also has the more subtle benefit of making decoding optional. That is, you can read the raw bytes with `ReadPrefix`, or skip straight to decoding them with `ReadObject`. Without the length prefix, the only way to get the raw bytes would be to decode into the object and then re-marshal it. `VerifyObject` needs to hash those bytes in order to check the signature, so it would wind up doing gross extra work.

Safe networking code requires length prefixes anyway, so if you drop the length prefix, you get this:
```go
err := encoding.WriteObject(conn, foo)
err = encoding.WritePrefix(conn, crypto.SignObject(bar))
var obj Object
bytes, err := encoding.ReadPrefix(conn)
err = crypto.VerifyObject(bytes, &obj)
```
Instead of:
```go
err := encoding.WriteObject(conn, foo)
err = crypto.WriteSignedObject(conn, bar)
var obj Object
err = crypto.ReadSignedObject(conn, &obj)
```

I'm fine with putting the signature at the beginning, especially if that's what other libraries do.